apisix/discovery/consul_kv.lua

```diff
@@ -0,0 +1,416 @@
+            enum = {"long", "short"},
+            default = "long"
+        },
+        prefix = {type = "string", default = "upstreams"},
+        weight = {type = "integer", minimum = 0, default = 1},
+        timeout = {
+            type = "object",
+            properties = {
+                connect = {type = "integer", minimum = 1, default = 2000},
+                send = {type = "integer", minimum = 1, default = 2000},
+                read = {type = "integer", minimum = 1, default = 2000}
+            },
+            default = {
+                connect = 2000,
+                send = 2000,
+                read = 2000
+            }
+        },
+        skip_keys = {
+            type = "array",
+            minItems = 1,
+            items = {
+                type = "string",
+            }
+        },
+        default_server = {
+            type = "object",
+            properties = {
+                host = {type = "string"},
+                port = {type = "integer"},
+                metadata = {
+                    type = "object",
+                    properties = {
+                        fail_timeout = {type = "integer", default = 1},
+                        weigth = {type = "integer", default = 1},
+                        max_fails = {type = "integer", default = 1}
+                    }
+                },
+                default = {
+                    connect = 1,
+                    send = 1,
+                    read = 1
+                }
+            }
+        }
+    },
+
+    required = {"servers"}
+}
+
+local _M = {
+    version = 0.3,
+}
+
+
+local function discovery_consul_callback(data, event, source, pid)
+    applications = data
+    log.notice("update local variable application, event is: ", event,
+        "source: ", source, "server pid:", pid,
+        ", application: ", core.json.encode(applications, true))
+end
+
+
+function _M.all_nodes()
+    return applications
+end
+
+
+function _M.nodes(service_name)
+    if not applications then
+        log.error("application is nil, failed to fetch nodes for : ", service_name)
+        return
+    end
+
+    local resp_list = applications[service_name]
+
+    if not resp_list then
+        log.error("fetch nodes failed by ", service_name, ", return default service")
+        return default_service and {default_service}
+    end
+
+    log.info("process id: ", ngx.worker.id(), ", applications[", service_name, "] = ",
+        core.json.delay_encode(resp_list, true))
+
+    return resp_list
+end
+
+
+local function parse_instance(node, server_name_prefix)
+    local cj = cjson
+    local key = node.Key
+    local pr = default_prefix_rule
+
+    if key == cj.null or not key or #key == 0 then
+        log.error("consul_key_empty, server_name_prefix: ", server_name_prefix,
+            ", node: ", core.json.delay_encode(node, true))
+        return false
+    end
+
+    local host, port, sn
+    for sn0, host0, port0 in string_gmatch(key, pr) do
```

Comment:
Could you use `ngx.re.gmatch`, which is much faster?

Reply:
Y